posts/MarkoB/2018-01-04-adding-new-constraint-to-engine.adoc

```diff
@@ -0,0 +1,488 @@
+
+An example of such test for the `@ISBN` constraint can be as follows:
+
+```java
+public class ISBNConstrainedTest extends AbstractConstrainedTest {
+
+    @Test
+    public void testISBN() {
+        Foo foo = new Foo( "978-1-56619-909-4" );
+        Set<ConstraintViolation<Foo>> violations = validator.validate( foo );
+        assertNoViolations( violations );
+    }
+
+    @Test
+    public void testISBNInvalid() {
+        Foo foo = new Foo( "5412-3456-7890" );
+        Set<ConstraintViolation<Foo>> violations = validator.validate( foo );
+        assertThat( violations ).containsOnlyViolations(
+                violationOf( ISBN.class ).withMessage( "invalid ISBN number" )
+        );
+    }
+
+    private static class Foo {
+        @ISBN
+        private final String number;
+
+        public Foo(String number) {
+            this.number = number;
+        }
+    }
+}
+```
+
+To make assertions on constraint violations, the `ConstraintViolationAssert` assertion class
+is used. It includes `ConstraintViolationAssert#assertNoViolations()` which will check that the
+passed set of constraint violations is empty.
+
+It also has a `ConstraintViolationAssert#assertThat()` method which receives a set of violations
+and returns a `ConstraintViolationSetAssert` which provides a rich API to perform assertions on violations.
+
+For the purposes of the `ConstrainedTest`, it is enough to check that the expected violations are present
+with the expected message (see `ISBNConstrainedTest#testISBNInvalid()` above).
+
+NOTE: Preferably, a `@TestForIssue` annotation should be added to all tests written for a
+constraint. This annotation can either be applied to a test method or to a test class.
+It has just one parameter - `jiraKey` which helps to link back a test to a corresponding
+JIRA ticket (for example `@TestForIssue(jiraKey = "HV-{number}")` where `{number}` is the
+number of the corresponding JIRA ticket).
+
+== Step 2 - Add programmatic definition
+
+**Commit:** https://github.com/hibernate/hibernate-validator/commit/91c4ed35b04c9a6a98ba9038aee514f94618fc6b
+
+The second main step is to add a programmatic definition for a new constraint. The definition
+itself is added to the `org.hibernate.validator.cfg.defs` package.
+It should extend `ConstraintDef` and provide methods which allow to specify all the
+constraint-specific attributes:
+
+```java
+public class ISBNDef extends ConstraintDef<ISBNDef, ISBN> {
+
+    public ISBNDef() {
+        super( ISBN.class );
+    }
+
+    public ISBNDef type(ISBN.Type type) {
+        addParameter( "type", type );
+        return this;
+    }
+}
+```
+
+The names of the methods in `YourConstraintDef` should match the names of the corresponding annotation
+attributes.
+The methods should also allow chaining so that constraint definition can be initialized
+in one go, hence they should return `this`.
+
+Importantly, these programmatic definitions should provide a fluent API and, in more complicated
```

Comment:
@marko-bekhta I think we should remove this paragraph. It makes the thing more complicated than it should be. And if someone wants to contribute a complicated constraint, we will be here to help and point him/her in the right direction.

WDYT?